### ISSUE TITLE
[FSDP] High CPU RAM usage with FSDP sharded checkpointing

### ISSUE THREAD

**Issue by konstantinjdobler** [pytorch/pytorch#118520]
### 🐛 Describe the bug

When using FSDP sharded checkpointing, I am seeing unusually high CPU RAM usage. Specifically, let's say I am training a 7B parameter model and checkpoints are saved in `float32` (so the optimizer state dict would be roughly 28GB). Then, during:
```python
optim_sd = get_optimizer_state_dict(model, optimizer, options=StateDictOptions(cpu_offload=True, full_state_dict=False))
```
it seems that **for each device**, CPU RAM according to the full state dict (so `world_size` * 28GB) is allocated (which leads to CPU OOM in my case), instead of only CPU RAM for the shard on the respective device. Is this expected behavior? My reading of https://github.com/pytorch/pytorch/issues/117476#issuecomment-1894590602 implies this might be the case.

My current workaround is to instead use `full_state_dict=True` for saving and loading when operating in environments contained by CPU RAM. Is there a better way?

cc: @awgu 

### Versions

```
Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.3
[pip3] pytorch-lightning==2.1.3
[pip3] torch==2.3.0.dev20240117
[pip3] torchmetrics==1.3.0
[pip3] triton==2.2.0
```

cc @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @penguinwu @kwen2501 @LucasLLC

**Comment by awgu** [comment]
cc: @fegin @wz337 @LucasLLC

**Comment by fegin** [comment]
@konstantinjdobler No, by design, we only call `to("CPU")` if `full_state_dict` is False. In such a case, if the tensor is a DTensor or ShardedTensor, the `to("CPU")` should only move the local shard to the CPU memory. I verified the behavior with the latest PyTorch code. Can you confirm that the `state_dict` you get contains either DTensor or ShardedTensor?

**Comment by konstantinjdobler** [comment]
Apologies for not interacting, I got sidetracked on work. I cannot reproduce the issue with the latest `torch`.
